10-JavaScript/classes.md

```diff
@@ -0,0 +1,247 @@
+
+```javascript
+class Book {
+  constructor(title, author) {
+    console.log("Constructor Ran!");
+    this.title = title;
+    this.author = author;
+  }
+}
+
+let poodr = new Book("Practical Object Oriented Programming in Ruby", "Metz");
+
+> Constructor Ran!
+```
+
+**Exercise**:  With a fellow classmate create an `Animal` class similar to `Book` above.  The `constructor` should take a `sound` parameter and save it in an instance variable.  Then create an instance of the class.  Verify that it works.
+
+### Instance Methods
+
+Instance methods are created like a constructors with their own names.  
+
+```javascript
+const SALESTAX = 0.08;
+
+class Book {
+  constructor(title, author, price) {
+    this.title = title;
+    this.author = author;
+    this.price = price;
+  }
+
+  totalPrice() {
+    return this.price * (1 + SALESTAX);
+  }
+}
+
+let adaBook = new Book("Practical Object Oriented Programming in Ruby", "Metz", 100);
+console.log(adaBook.totalPrice());
+
+> 108
+```
+
+**Exercise**:  For our `Animal` class, create a `speak` method which will print the sound to the console.  Try it out to verify that the method is working.
+
+### Getter & Setter Methods
+
+Getter methods are methods which are run when a property with the same name is accessed.  Similarly a Setter method is called when there is an attempt to change a property.  
+
+```javascript
+class Book {
+  constructor(title, author) {
+    console.log("Constructor Run!");
+    this._title = title;
+    this._author = author;
+  }
+
+  // getter methods
+  get author() {
+    return this._author;
+  }
+
+  // setter method
+  set author(newAuthor) {
+    if (typeof newAuthor === "string") {
+      this._author = newAuthor;
+    }
+  }
+
+}
+let poodr = new Book("Practical Object Oriented Programming in Ruby", "Metz");
+
+// setter method
+poodr.author = "Chris";
+
+console.log(`${poodr.author}`); // getter method
+> Chris
+```
+
+Notice that the properties are all named __\_propertyName__.  This is a typical naming convention for instance variables you want to keep private.  It's a convention that signals an intent to prevent access outside the class as nothing is **really** private in JavaScript.  Properties cannot have the same names as our getter and setter methods as doing so will result in infinite recursion.  
+
+**Question**:  Why would naming the property the same as the getter or setter method result in infinite recursion?
+
+**Exercise**:  Create getter and setter methods for the `Animal`'s `sound`' so that you can do:  `myAnimal.sound = "Ruff"` and `console.log(myAnimal.sound)`.
+
+### Static Methods
+
+Static methods are equivalent to class methods in Ruby.  They are attached to the class rather than instances of the class.
+
+
+```javascript
+class Book {
+  constructor(title, author, price) {
+    this.title = title;
```

Comment:
Do we want to continue using the getter/setters and using the underscore on these variables?

Reply:
I can, I was trying to keep the code snippets small to illustrate points.

Reply:
That's fine - we can just add a note instead so they know why we stopped using it.